src/random_ray/random_ray.cpp

```diff
@@ -0,0 +1,288 @@
+
+    // Aquire lock for source region
+    random_ray::lock[source_region].lock();
+
+    // Accumulate delta psi into new estimate of source region flux for
+    // this iteration
+    for (int e = 0; e < negroups; e++) {
+      random_ray::scalar_flux_new[source_element + e] += delta_psi_[e];
+    }
+
+    // If the source region hasn't been hit yet this iteration,
+    // indicate that it now has
+    if (random_ray::was_hit[source_region] == 0) {
+      random_ray::was_hit[source_region] = 1;
+    }
+
+    // Accomulate volume (ray distance) into this iteration's estimate
+    // of the source region's volume
+    random_ray::volume[source_region] += distance;
+
+    // Tally valid position inside the source region (e.g., midpoint of
+    // the ray) if not done already
+    if (!random_ray::position_recorded[source_region]) {
+      Position midpoint = r() + u() * (distance / 2.0);
+      random_ray::position[source_region] = midpoint;
+      random_ray::position_recorded[source_region] = 1;
+    }
+
+    // Release lock
+    random_ray::lock[source_region].unlock();
+  }
+}
+
+void RandomRay::initialize_ray(uint64_t ray_id, int sampling_source)
+{
+  // Reset particle event counter
+  n_event() = 0;
+
+  if (settings::random_ray_distance_inactive <= 0.0)
+    is_active_ = true;
+  else
+    is_active_ = false;
+
+  wgt() = 1.0;
+
+  // set identifier for particle
+  id() = simulation::work_index[mpi::rank] + ray_id;
+
+  // set random number seed
+  int64_t particle_seed =
+    (simulation::current_batch - 1) * settings::n_particles + id();
+  init_particle_seeds(particle_seed, seeds());
+  stream() = STREAM_TRACKING;
+
+  // Sample from ray source distribution
+  SourceSite site {
+    model::external_sources[sampling_source]->sample(current_seed())};
+  site.E = lower_bound_index(
+    data::mg.rev_energy_bins_.begin(), data::mg.rev_energy_bins_.end(), site.E);
+  site.E = data::mg.num_energy_groups_ - site.E - 1.;
+  from_source(&site);
+
+  // Locate ray
+  if (lowest_coord().cell == C_NONE) {
+    if (!exhaustive_find_cell(*this)) {
+      this->mark_as_lost(
+        "Could not find the cell containing particle " + std::to_string(id()));
+    }
+
+    // Set birth cell attribute
+    if (cell_born() == C_NONE)
+      cell_born() = lowest_coord().cell;
+  }
+
+  // Initialize ray's starting angular flux to starting location's isotropic
+  // source
+  int negroups = data::mg.num_energy_groups_;
+  int i_cell = lowest_coord().cell;
+  int64_t source_region_idx =
+    random_ray::source_region_offsets[i_cell] + cell_instance();
+
+  for (int e = 0; e < negroups; e++) {
+    angular_flux_[e] = random_ray::source[source_region_idx * negroups + e];
+  }
+}
+
+//==============================================================================
+// Non-method functions
+//==============================================================================
+
+// returns 1 - exp(-tau)
+// Equivalent to -(_expm1f(-tau)), but is stable
```

Comment:
Replace `Equivalent` with `Approximates`. Also pretty sure the original is stable. This one is just faster.